### PR TITLE
Cloud run tasks (post-plan only) CLI integration

### PR DESCRIPTION
**Pull request by uturunku1**
Integration with Terraform Cloud Run Tasks (post-plan stage). Run tasks is a beta feature implemented on Terraform Cloud that will display results after a plan like so:

Screen Shot 2021-12-10 at 11 30 52 AM

These results are currently not being displayed or consumed by Terraform CLI, so with this PR (as well as some changes made to go-tfe https://github.com/hashicorp/go-tfe/pull/286), is implementing the necessary changes to display the output of Run Tasks' result in the CLI:

Screen Shot 2021-12-10 at 11 14 50 AM

### IMPLEMENTATION

All changes in this PR are being applied through the `op.plan` because we want the user to see results of Run Tasks after a plan has found changes to infrastructure and right before the apply operation (so what we call the `pre-apply` stage).

These changes are also scope to the `cloud` package. This is to be determined, but we think that the CLI will not offer support for Run Tasks prior to version 1.2

This PR also adds an abstraction when interacting with Cloud Integrations such as Run Tasks, Policy Checks and Cost estimation. The abstraction is not yet applied to Policy Checks and Cost Estimation. But if we were to use it for all three Cloud Integrations, it would simplify and unify output during the pre-apply stage.

### TESTING

1. Ensure your organization has access to the Run Tasks beta. You may have to request access for staging and prod. However, if you are running Atlas on dev mode, you'll have Run Task feature enabled by default.
2. Follow these steps to add run tasks to the workspace you'll be testing with: https://www.terraform.io/docs/cloud/workspaces/run-tasks.html
3. We have a go server running on a EC2 instance available for test event hooks. This are the URLs you can add to your run tasks:

http://54.167.177.151/success
http://54.167.177.151/failed
http://54.167.177.151/success?timeout=630 to force a timeout. But you can change the value of timeout to as many seconds you want the server to wait before firing off the webhook back to TFC

4. Create some config using a `cloud` config

```
terraform {
  cloud {
    hostname = "app.staging.terraform.io"
    organization = "test"
    workspaces {
      name = "project"
    }
  }
}

variable "replicas" {
  type = number
  default = 7
}

output "hello_there" {
  value = var.replicas
}

```

5. Using a testing workspace, run `./terraform apply` to see Run Tasks output right after Cost Estimation output. If you have no tasks added to your workspace in the platform, you should see no output related to Run Tasks.

### Screenshots

Task results
Screen Shot 2021-12-15 at 9 53 53 AM

No run tasks defined
Screen Shot 2021-12-15 at 9 59 56 AM

No infrastructure changes (unreachable tasks)
Screen Shot 2021-12-15 at 10 00 05 AM

NOTE: Once this PR has being approved, we will squash the commits.

### CODE REVIEW

File: internal/cloud/backend_plan.go
Comment (brandonc) [inline internal/cloud/backend_plan.go:357]: This needs to be addressed before merging

**Review by barrettclark — APPROVED** [review]
This all looks good to me. I left 1 nit. The PR also has a merge conflict that will need to be resolved.

File: internal/cloud/backend_runTasks.go
Comment (uturunku1) [reply internal/cloud/backend_runTasks.go:22]: Yes, I totally agree with this change. I guess I was trying to stick a bit to the "Go way" of naming variable, which usually is one letter or very few letters (like "p", "ret", "r", more on that topic: https://blog.devgenius.io/write-better-go-code-by-keeping-your-variable-names-short-13f404ac515c), which I actually find a bit annoying because, like you, I prefer descriptive names. Also, I have noticed that this project does not care much for that type of naming conventions, so I think is best for me to stay away from single letters :) @barrettclark

Comment (barrettclark) [reply internal/cloud/backend_runTasks.go]: It's wild to me that the Go way would favor terse over descriptive variables. When future me drops into old code and see something like `xzz` I have to spend more time trying to remember what the heck that means and why. I will try to not swim against the current if our style guide favors brevity, but very much appreciate more verbose and meaningful names in the meantime. Thank you for making this change.

File: go.mod
Comment (nfagerlund) [inline go.mod:44]: Just a gut-check -- are we thinking to wait to merge this PR until the go-tfe changes have merged? Or will we treat this more like the run-up to 1.1.0, when we were building against a WIP branch for a while? 

If it's the latter, we should probably rebase the WIP branch and make sure we don't lose any features compared to what were previously at. It looks like with the current status, we might lose run variables?

**Review by nfagerlund — APPROVED** [review]
I did some manual testing, and it looks pretty sweet! One question about the status of the go-tfe work, but otherwise this seems good.

File: go.mod
Comment (brandonc) [reply go.mod:44]: We need to have a slightly better process I think for pre-release go-tfe. I think we are going to wait on the go-tfe merge and depend directly on a branch until the next release of terraform that includes run tasks, similar to what we did in the run up to 1.1. Reason being, we want to make sure the API doesn't shift before releasing in go-tfe. This shouldn't compile if run variables weren't in the base, so it must be based on a run variables pre-release. I take your point and will make sure to rebase this to v0.22.0.

**Comment by brandonc** [comment]
Updated go-tfe (using branch run-tasks-integration)

File: internal/cloud/backend_plan.go
Comment (chrisarcand) [inline internal/cloud/backend_plan.go:336]: This feels brittle to me. Although it's not ideal, the 'proper' way to do this would be to use the `TFP-API-Version` header information (exposed in the client [here](https://github.com/hashicorp/go-tfe/blob/73c3b410274fc2d42eeb60ded9157ae325484301/tfe.go#L280-L298)) to conditionally make a single request if the host (TFC or TFE) supports the necessary API version for it.

Comment (brandonc) [reply internal/cloud/backend_plan.go:336]: That is my intention! Once this is merged I can supplant the next version with run tasks. Feels bad though because I know run tasks will squat on that version for a long time. https://github.com/hashicorp/atlas/pull/11572

Comment (uturunku1) [reply internal/cloud/backend_plan.go:336]: If we hit line 335, and TFE or TFC have Run Tasks defined for that workspace, then the CLI would not display anything about run tasks' results but the platform would still show run tasks' results, which is the discrepancy/problem we were trying to solve to begin with. I wasn't sure if that's the behavior we want, but I want to bring that up just in case. I understand these changes are temporary. @brandonc @chrisarcand

Comment (brandonc) [reply internal/cloud/backend_plan.go:336]: If we hit line 335, run tasks won't be defined for the workspace because we'll be working against a version of TFE that doesn't support them -- that's what "invalid include parameter" would indicate... But it's not a great feature detection strategy and I'd like to rely on the API version instead.

File: internal/cloud/backend_plan.go
Comment (brandonc) [inline internal/cloud/backend_plan.go:349]:
```suggestion
	// Await post-plan run tasks
```

😄

Comment (sebasslash) [reply internal/cloud/backend_plan.go]: Silly me!

**Pull request merged**
